src/Content/Conversation/Repository/UserDefinedChannel.php

```diff
@@ -140,40 +150,174 @@ public function save(Entity\UserDefinedChannel $Channel): Entity\UserDefinedChan
 		return $Channel;
 	}
 
+	private function isValid(string $searchtext): bool
+	{
+		if ($searchtext == '') {
+			return true;
+		}
+
+		return $this->db->select('check-full-text-search', [], ["`pid` = ? AND MATCH (`searchtext`) AGAINST (? IN BOOLEAN MODE)", getmypid(), $this->escapeKeywords($searchtext)]) !== false;
+	}
+
 	/**
-	 * Checks, if one of the user defined channels matches with the given search text
-	 * @todo To increase the performance, this functionality should be replaced with a single SQL call.
+	 * Checks, if one of the user defined channels matches with the given search text or languages
 	 *
 	 * @param string $searchtext
 	 * @param string $language
 	 * @return boolean
 	 */
 	public function match(string $searchtext, string $language): bool
 	{
+		$users = $this->db->selectToArray('user', ['uid'], $this->getUserCondition());
+		if (empty($users)) {
+			return [];
+		}
+
+		$uids = array_column($users, 'uid');
+
+		$usercondition = ['uid' => $uids];
+		$condition = DBA::mergeConditions($usercondition, ["`languages` != ? AND `include-tags` = ? AND `full-text-search` = ? AND `circle` = ?", '', '', '', 0]);
+		foreach ($this->select($condition) as $channel) {
+			if (!empty($channel->languages) && in_array($language, $channel->languages)) {
+				return true;
+			}
+		}
+
+		$search = '';
+		$condition = DBA::mergeConditions($usercondition, ["`full-text-search` != ? AND `circle` = ? AND `valid`", '', 0]);
+		foreach ($this->select($condition) as $channel) {
+			$search .= '(' . $channel->fullTextSearch . ') ';
+		}
+
+		$this->db->insert('check-full-text-search', ['pid' => getmypid(), 'searchtext' => $searchtext], Database::INSERT_UPDATE);
+		$result = $this->inFulltext($search);
+		$this->db->delete('check-full-text-search', ['pid' => getmypid()]);
```

Comment:
Is all this necessary after you changed the validity test? Is the `check-full-text-search` table necessary?

Reply:
Yes. Here we don't check the validity. Here we check, if we have got a match.

Reply:
A match between what exactly? I'm very confused by this code.

Reply:
This specific function checks if there are any user defined channels that would match to the given search text. This function is called at two places. First it is called for posts that arrived via relay. We don't keep all posts there if not explicitly selected. So we check if the post meets some expectations. One is, that it contains hashtags that users subscribed to. And another check is, that there are user defined channels that would select this post.

Reply:
So we can perform the search directly on the channels table, what's the use of `check-full-text-search` besides checking for search string validity that we got rid of?

Reply:
No. We execute this query before we consider to even store the content.

Reply:
I'm thoroughly lost by now. where is the search string defined? What it is supposed to match against? Why do we need the `check-full-text-search` table?

Reply:
We need to temporarily store the text of the message to a real table (no temporary tables are allowed) to perform the search. That `check-full-text-search` has got the process id as primary key, since there can be a lot of parallel access to that table.

Reply:
Ah, thanks, I understand now. Isn't it easier to perform the search in a regular expression instead? Seems less complicated than inserting a row and then immediately removing it from the database.

Reply:
You can use complex search patterns: https://mariadb.com/kb/en/full-text-index-overview/#in-boolean-mode

This means that I would need to create some parser that emulates that search, just to avoid to insert some data into some table.